File: workspace/2023/2023-03-18/numerical-sequence.prolog/fizz-buzz.pl

%
% N 番目の FizzBuzz.
%

fizz_buzz(N, 'FizzBuzz') :-
    N mod 15 =:= 0,
    !.

fizz_buzz(N, 'Buzz') :-
    N mod 5 =:= 0,
    !.

fizz_buzz(N, 'Fizz') :-
    N mod 3 =:= 0,
    !.

fizz_buzz(N, N).

%
% Min 番目から Max 番目までの FizzBuzz.
%

fizz_buzz_values(Min, Max, []) :-
    Min > Max.

fizz_buzz_values(Min, Max, [Fmin|Xs]) :-
    Min =< Max,
    fizz_buzz(Min, Fmin),
    is(NextMin, Min + 1),
    fizz_buzz_values(NextMin, Max, Xs).

File: workspace/2025/2025-06-16/numerical-sequence.forth/fizz-buzz.fs

: fizz_buzz ( n -- c-addr u )
    dup 15 mod 0 = if
        drop s" FizzBuzz"
    else
        dup 5 mod 0 = if
            drop s" Buzz"
        else
            dup 3 mod 0 = if
                drop s" Fizz"
            else
                s>d <# #s #>
            endif
        endif
    endif
;

: print_fizz_buzz_values_range recursive ( max min -- )
    over over < if
        drop drop
    else
        dup fizz_buzz
        3 pick 3 pick <> if s"  " s+ endif type
        1 + print_fizz_buzz_values_range
    endif
;

: print_fizz_buzz_values ( n -- )
    s\" fizz buzz values:\n" type
    1 print_fizz_buzz_values_range
    s\" \n" type
;

